Feature/deferred ops

idodeclare opened the pull request:
Hello,

Please consider for integration this patch with some improvements in error handling and changes to reduce the time interval where the Lucene database is inconsistent with the xref file data and to speed up deletions for issue #1908.

1) Add `PendingFileCompleter` and use in `IndexDatabase` to queue up file deletions and file renamings to be executed during the Lucene 2-phase commit. `IndexDatabase` is changed to write to in-progress files (i.e., with an "._org_opengrok" extension) while xref-ing and to queue up the moves to final names (without the pending extension) during the 2-phase commit. This minimizes the time interval of inconsistency between the xref data and the committed Lucene data.
2. Do not run Lucene `commit()` on each file deletion, but instead once as the last step of `IndexDatabase.update()` so that Lucene and `PendingFileCompleter` can be well-coordinated.
3) Do not swallow `IndexWriter` exceptions for `prepareCommit()`, `commit()`, or `close()` in either the `IndexDatabase` `update()` or `optimize()` methods.
3) Support `NativeFSLockFactory`, and support the `--lock` switch via the OpenGrok script (`on|off|native|simple` with the default still `off`, and `on` as an alias for `simple`). `Configuration` is revised to support the new lock mode, with the existing `boolean usingLuceneLocking` preserved but `@Deprecated` to be retired at some future point — i.e., after a sufficient upgrade time interval is deemed to have past.
4) Ensure a meaningful non-zero exit code if a non-fatal error occurs in the OpenGrok script.

Thank you.

<!--
Thank you for proposing a contribution to the {OpenGrok project. In order to accept changes from the "outside world", all contributors should "sign" the [Oracle Contributor Agreement](www.oracle.com/technetwork/community/oca-486395.html) . 
OCA basically means that you won't be sueing Oracle over code you donate to {OpenGrok project (and similar way, that Oracle won't sue you over your patch :-D ).
If you have already signed OCA or are from Oracle, then ignore this message, you just need to sign once for all patches to {OpenGrok.
Alternative is to provide a written acceptance of OCA line into the comment of specific pull request (and ideally sign, scan and mail back OCA to Oracle in parallel), but this simple acceptance is only valid for single pull request.
-->

vladak commented:
It'd be nice to have some measurements to see if this really helps #1908.

tarzanek commented:
@tuxillo will you be able to help here with testing ?
(build this branch, set it up and try in your scenario?)

tarzanek commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -93,9 +96,17 @@
+     * draft of a policy value used to flush -- not commit -- deleted Lucene
+     * items periodically.
+     */
+    private static final int MAX_BUFFERED_DELETE_TERMS = 200;
```
would it make sense to expose this as tunable?

idodeclare replied:
I would say no. I do not imagine anyone would be able to definitively trace any performance issue to this setting — and compared to the baseline behavior of OpenGrok where it executes `commit()` on each deletion of document, any non-zero, but "reasonable" (i.e., not too far from zero) value here seems OK.

tarzanek commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -1295,17 +1316,59 @@ private Writer getXrefWriter(FileAnalyzer fa, String path) throws IOException {
         return null;
     }
+
+    LockFactory pickLockFactory(RuntimeEnvironment env) {
```
lockpick factory :-D

tuxillo commented:
Hi,

Yeah, will give it a try as soon as the current re-indexing is done. But I am not sure if I can provide exactly the same scenario as I have it right now since the next repository update might have fewer files than this previous one.

Although I cannot do an exact measurement I think any improvement would be noticeable because right now the deletion rate is around 2-3 files per second.

Regards,
Antonio Huete

vladak commented:
w.r.t. index inconsistency - keep in mind that the webapp will switch to the next version of the index only after it is explicitly told so via `maybeRefreshIndexSearchers()` that is sent at the end of the indexing via `RefreshMessage`. So even if the commits are delayed, this will not change much for the inconsistency window from the perspective of the user.

idodeclare commented:
> It'd be nice to have some measurements to see if this really helps #1908.

Well I can show it noticeably helps even a smallish repo.

Following is logging from a test with the existing logic for a 31MB-sized repo (Homebrew/brew) where I deleted 1008 files (`git rm -r Library/`) after having indexed first the repo:

> 2017-12-08 07:26:21.361-0600 INFO t20 DefaultIndexChangedListener.fileRemove: Remove file:/brew/Library/.editorconfig
> 2017-12-08 07:26:21.391-0600 FINER t20 DefaultIndexChangedListener.fileRemoved: Removed file:/brew/Library/.editorconfig
> ...
> 2017-12-08 07:26:27.151-0600 INFO t20 DefaultIndexChangedListener.fileRemove: Remove file:/brew/Library/README.md
> 2017-12-08 07:26:27.154-0600 FINER t20 DefaultIndexChangedListener.fileRemoved: Removed file:/brew/Library/README.md
> 

That is nearly six seconds (and 1008 extra `commit()`s).

Following is the same test with this patch where filesystem deletes and `commit()` are deferred:

> 2017-12-08 07:30:39.071-0600 INFO t20 DefaultIndexChangedListener.fileRemove: Remove file:/brew/Library/.editorconfig
> 2017-12-08 07:30:39.076-0600 FINER t20 DefaultIndexChangedListener.fileRemoved: Removed file:/brew/Library/.editorconfig
> ...
> 2017-12-08 07:30:40.240-0600 INFO t20 DefaultIndexChangedListener.fileRemove: Remove file:/brew/Library/README.md
> 2017-12-08 07:30:40.242-0600 FINER t20 DefaultIndexChangedListener.fileRemoved: Removed file:/brew/Library/README.md
> 2017-12-08 07:30:40.390-0600 FINER t25 PendingFileCompleter.doDelete: Deleted obsolete file: /private/var/opengrok/data/xref/brew/Library/Homebrew/test/cask/dsl/uninstall_preflight_spec.rb.gz
> ...
> 2017-12-08 07:30:40.705-0600 FINE t20 IndexDatabase.finishWriting: completed 1,008 file(s)
> 

That is just over a second, saving five seconds for this smallish repo — or ~80% less time. The `PendingFileCompleter` logging shows that the filesystem deletes take only 300 ms, so the vast bulk of the savings is in deferred `commit()`s.

The tests above show an 80% savings. As `commit()` becomes more expensive, the savings grows and grows. @tuxillo's logging for his 110GB repo (70GB Lucene) shows more than half-a-second per file (versus ~15 ms for stage-one of my tests). Even assuming that the per-document cost of Lucene `deleteDocument` at that size is 1000% more expensive than my tests, my worksheet shows the time saved in deferring is 96%.

tuxillo commented:
Hi,

The reindexing has took almost 10 days this time. I'm up for testing your patch but how should I approach it? Currently I'm using 1.1-rc16, should I just use 'master' branch directly and apply these changes to it?

Regards,
Antonio Huete

tarzanek commented:
ideally use just this branch, clone it and build it so you test just this change and not others that get added in parallel

tarzanek commented:
is 10 days expected?
For me anything > 8h must be explained by something - e.g. too big codebase, too low tuning, git rename history enabled or bugs ( leaks, problematic I/O, blocking code)
Do we have an explanation for 10 days? (feel free to throw your sizes at us (codebase, server size used, allocations) and we can suggest some tuning)

tuxillo commented:
Hi,

Check #1908 , all the details about the sizes and env variables we use are there.
Let me know if we can provide further details.

Regards,
Antonio Huete

tarzanek commented:
this definitely shouldn't take 10 days on such a strong HW, so this kinda looks like GIT renames are in place and slowing the indexing down ... (but then recent builds should have git renames turned off ... )
you said that disk is idle ... can you trace the process somehow to see what is it doing for most of the time?

tarzanek commented:
I think @vladak handles even bigger codebase and index is much faster ...
@tuxillo how long did the initial index take ?

tuxillo commented:
The initial indexing took around 80h, it's the reindexing which takes a lot of time because of the slowness in the file removal from the big index.

tarzanek commented:
so with this patch it is 10 days and it used to be more?
(just trying to get the problem definition more precise, hence the silly Qs)

tuxillo commented:
Sorry, I'm not being clear. I have two 'data' and two 'src' directories which are the same. When I am serving one, I am reindexing, just like it's recommended in the documentation. Now, when I reindexed 'src1' it took almost 8 days to do it (a lot of files were removed from the biggest index) and when I reindexed 'src2' it took nearly 10 days.
Both operations were done with 1.1-rc16.

Now I'm about to try the patch and I'll report back the timing to see the difference.

idodeclare commented:
Seeing folks write here about needing hours to index, I've pushed a commit to parallelize `IndexDatabase.update()`  — formerly, as has been documented, a single thread would do the indexing within any `IndexDatabase` instance.

Now a `ForkJoinPool` will be shared among `IndexDatabase` instances to enforce global limits (see new env var `OPENGROK_PARALLELISM` in the `OpenGrok` script using the existing `Indexer` `--threads` switch) but to allow even a single `IndexDatabase.update()` (or e.g. the last instance of a large set of projects) to be parallelized too.

This should also resolve issue #460, issue #802, and issue #1451 .

tuxillo commented:
I've conducted an initial test with 1.1-rc16 vs master w/ deferred_ops commits and it shows a huge improvement as far as I can see:

A) 1.1-rc16 Total time: 14:20:24

Added files: 42598
Removed files: 62205

B) master (dec 14th) with deferred ops commits: Total time: 10:51:26 

Added files: 42788
Removed files: 62422

In this case, near the end of the reindexing there were about 50 exctags processes hanging which I had to kill so that the process finished. The process started at 19:44 and it finished around 22:40 when the exctags were hanging and had to be killed so I'd say the real reindexing time is around ~3h.

--- 

I've put up all the logs here so you can have a look for yourselves. I've also included a vmstat log which is started at the beginning of the reindexing process (although it contains records after the reindexing process was long done)

Logs: https://monster.dragonflybsd.org/~tuxillo/grok_bench/1/

Next step is doing a lot of changes to the big repo and doing a reindexing with B) unless you state otherwise. Let me know.

idodeclare commented:
@tuxillo , those are encouraging results.

From the logs, I see some Ctags instances' processes were hung up very near to their creation and first writing. One clue I found at https://docs.oracle.com/javase/7/docs/api/java/lang/Process.html is that "failure to promptly ... read the output stream of the subprocess may cause the subprocess to block, or even deadlock." I have pushed a revision so that the read from the subprocess's `stdout` happens on a separate thread from the write to its `stdin` in hopes that it cures the hangups.

When convenient, please do pull the revision and continue your next phase of testing.

tarzanek commented:
I hope the hanging ctags are not caused by our change in jflex files and limiting tokens still works  ...

tarzanek commented:
(I have a sample xml file that I can use to test the hanging ... I need to make it a test, but it has high impact on tests, since it takes a long while for parser to parse it ...)

idodeclare commented:
> I hope the hanging ctags are not caused by our change in jflex files and limiting tokens still works ...

I see that `Ctags doCtags()` is run synchronously before either the symbol tokenizer or the xrefer is used by any `PlainAnalyzer` subclass, so it doesn't seem as if any jflex change could have that effect.

Also, @tuxillo's killing the ctags processes cleared the blockage, with the exceptions caught and `IndexDatabase` skipping the files because "because the analyzer didn't understand [them]."

idodeclare commented:
Oops should be: "ensure Ctags does _not_ block indefinitely"

tuxillo commented:
Hi,

The last check was with master as of:
0d47f77 -  Merge pull request #1953 from idodeclare/feature/powershell_tests

And with this PR changes up until:

932fd09 - Read ctags output from a separate thread

The result is the same, exctags processes are hung. In fact I even went on holidays and when I got back the processes where still there. Here's the log:

    2017-12-29 15:05:29.009-0800 FINEST t564 PendingFileCompleter.doRename: Moved pending as file: /build3/grok/data1/xref/dports/lang/gcc8-devel/gcc-8-20171126/gcc/testsuite/g++.dg/ipa/devirt-25.C.gz
    2017-12-29 15:05:29.010-0800 FINEST t564 PendingFileCompleter.doRename: Moved pending as file: /build3/grok/data1/xref/dports/security/gnupg/gnupg-2.2.3/build-aux/config.sub.gz
    2017-12-29 15:05:29.010-0800 FINEST t564 PendingFileCompleter.doRename: Moved pending as file: /build3/grok/data1/xref/dports/ftp/curl/curl-7.57.0/tests/data/test1306.gz
    2017-12-29 15:05:29.020-0800 FINE t221 PendingFileCompleter.complete: renamed 931,853 file(s)
    2017-12-29 15:05:29.021-0800 FINE t221 IndexDatabase.finishWriting: completed 1,898,361 file(s)
    2017-12-29 15:05:29.122-0800 INFO t221 IndexDatabase.optimize: Optimizing the index     
    2017-12-29 17:48:21.462-0800 INFO t221 IndexDatabase.optimize: done
    [...]
    2018-01-07 17:34:36.642-0800 FINE t1 Ctags.close: Destroying ctags command
    2018-01-07 17:34:36.645-0800 FINE t1 Ctags.close: Destroying ctags command
    2018-01-07 17:34:36.648-0800 FINE t1 Ctags.close: Destroying ctags command
    2018-01-07 17:34:36.652-0800 FINE t1 Ctags.close: Destroying ctags command
    2018-01-07 17:34:36.777-0800 INFO t1 Statistics.report: Done indexing data of all repositories (took 9 days 8:59:00)
    2018-01-07 17:34:36.782-0800 INFO t1 Indexer.writeConfigToFile: Writing configuration to /var/opengrok/etc/configuration.xml
    2018-01-07 17:34:39.110-0800 INFO t1 Indexer.writeConfigToFile: Done...
    2018-01-07 17:34:39.112-0800 INFO t1 Statistics.report: Total time: 9 days 9:27:30
    2018-01-07 17:34:51.820-0800 INFO t1 Statistics.report: Final Memory: 258M/14,955M


-----

Should I try with latest changes on this PR on top of latest master now?

Regards,
Antonio Huete

idodeclare commented:
@tuxillo, yes please do.

tuxillo commented:
Same result, a lot of exctags process hanging, I think waiting for input from a pipe or something similar. 
At 6:30 I had to kill them all so that the process could finish successfully.

    2018-01-11 06:31:52.159-0800 FINEST t1063 PendingFileCompleter.doRename: Moved pending as file: /build3/grok/data2/xref/dragonfly/sys/sys/poll.h.gz
    2018-01-11 06:31:52.159-0800 FINEST t1029 PendingFileCompleter.doRename: Moved pending as file: /build3/grok/data2/xref/dragonfly/sys/dev/raid/vinum/vinumdaemon.c.gz
    2018-01-11 06:31:52.159-0800 FINEST t1036 PendingFileCompleter.doRename: Moved pending as file: /build3/grok/data2/xref/dragonfly/usr.sbin/arp/arp.c.gz
    2018-01-11 06:31:52.161-0800 FINE t111 PendingFileCompleter.complete: renamed 311 file(s)
    2018-01-11 06:31:52.161-0800 FINE t111 IndexDatabase.finishWriting: completed 333 file(s)
    2018-01-11 06:33:29.016-0800 INFO t111 IndexDatabase.optimize: Optimizing the index ... 
    2018-01-11 06:40:30.785-0800 INFO t111 IndexDatabase.optimize: done
    2018-01-11 08:30:10.838-0800 INFO t117 IndexDatabase.optimize: done
    2018-01-11 08:30:12.562-0800 FINE t1 Ctags.close: Destroying ctags command
    2018-01-11 08:30:12.563-0800 FINE t1 Ctags.close: Destroying ctags command
    2018-01-11 08:30:12.563-0800 FINE t1 Ctags.close: Destroying ctags command
    2018-01-11 08:30:12.568-0800 INFO t1 Statistics.report: Done indexing data of all repositories (took 10:48:56)
    2018-01-11 08:30:12.568-0800 INFO t1 Indexer.writeConfigToFile: Writing configuration to /var/opengrok/etc/configuration.xml
    2018-01-11 08:30:13.423-0800 INFO t1 Indexer.writeConfigToFile: Done...
    2018-01-11 08:30:13.424-0800 INFO t1 Statistics.report: Total time: 11:16:37
    2018-01-11 08:30:17.887-0800 INFO t1 Statistics.report: Final Memory: 260M/14,755M

idodeclare commented:
Do you see warnings in the log: `Ctags futureTimeout executed`?

tuxillo commented:
The first one appears at:

    2018-01-11 01:04:31.155-0800 WARNING t186 Ctags.lambda$doCtags$1: Ctags futureTimeout executed.

The last one at:

    2018-01-11 03:35:38.510-0800 WARNING t189 Ctags.lambda$doCtags$1: Ctags futureTimeout executed.

A total of:

    # grep -c futureTimeout /tmp/opengrok0.0.log 
    227

idodeclare commented:
@tuxillo, I pushed a change to `destroyForcibly()` a hanging ctags process while still signalling the client thread so that retry happens.

tuxillo commented:
Ok, I'm going to give it try and I'll report back. I'll be using latest master plus the latest commit you just added.

tuxillo commented:
Still the same, exctags process waiting to read from a pipe.

Full log: https://monster.dragonflybsd.org/~tuxillo/grok_bench/2/opengrok0.0.log
ps aux for the exctags processes: https://monster.dragonflybsd.org/~tuxillo/grok_bench/2/exctags.log

idodeclare commented:
@tuxillo , thank you for continued help in testing.

From your `opengrok0.0.log`, I see on line 4,397 that thread 265 began starting a ctags instance, but then there is nothing further from that thread until line 9,643,496 where the exception related to your killing its process is logged. There is no related `futureTimeout` which indicates to me that `ProcessBuilder` hung up before the timeout could even be set.

I have pushed a change to work around such a condition.

(N.b. there are a total of four `futureTimeout` occurrences in that log—all unrelated to thread 265 and all proceeding normally. I have extended the timeout by another thirty seconds.)

tuxillo commented:
Same thing, I had to kill the exctags processes once again. The difference this time is that I had to kill them a few times because they were getting called after the hanging ones were killed.

Log: https://monster.dragonflybsd.org/~tuxillo/grok_bench/4/opengrok0.0.log

tarzanek commented:
wondering if we should split this to have a fix for the deletion first and focus on ctags pooling in other PR, hmm ?

idodeclare commented:
@tuxillo , @tarzanek , my research indicates that the problem may be that a JRE's implementation of `ProcessBuilder` could be the limiting factor here.

@tuxillo , OpenGrok executes long-running `ctags` processes that accept intermittent work (i.e., a filename) and then produce tags to be collected by OpenGrok.

With this branch, the default parallelism is to use all available processors. On @tuxillo's Opteron 6168 48 CPU and openjdk version "1.8.0_144" on DragonFly BSD, I wonder if that is too many simultaneous (Java-started) long-running processes if e.g. the JRE is `fork()`ing a process instead of `posix_spawn()`ing it.

@tuxillo , please would you try dialing down the parallelism — e.g. by setting `OPENGROK_PARALLELISM` to something much lower for the `OpenGrok` script: `export OPENGROK_PARALLELISM=8`.

tarzanek commented:
that could be, also it could be it's something BSD specific

tuxillo commented:
@idodeclare Okay, I'll lower the parallelism to 8 and try again. Anything else I should take in account?

If there is a test suite or something I could run to verify the platform has no problems for OpenGrok's specific case I can do that.

idodeclare commented:
> Anything else I should take in account? If there is a test suite or something I could run to verify the platform has no problems for OpenGrok's specific case I can do that.

@tuxillo, all the testing so far shows that `ProcessBuilder start()` is hanging up intermittently. Unfortunately I haven't found much recent information about such problems, and I cannot reproduce it on my much more modest hardware running Mac OS X or Linux.

I scanned through [UNIXProcess.java.bsd](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/solaris/classes/java/lang/UNIXProcess.java.bsd) and and [UNIXProcess_md.c](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/solaris/native/java/lang/UNIXProcess_md.c), and it does seem that BSD ports of jdk8 will use the more desirable `posix_spawn` by default.

Alas, I'm not aware of any test suite regarding this.

I am hopeful to learn whether dialing down`OPENGROK_PARALLELISM` produces any different result.

As a very long shot to see if there might be a `ProcessBuilder` deadlock on your hardware when multiple threads use `start()` simultaneously for `exctags`, another experiment to try might be to serialize the `initialize()` calls in Ctags.java (line 344 in 736e80ebc71) by synchronizing on a private class locking Object (`static final`).

tuxillo commented:
Hi,

It seems that setting OPENGROK_PARALLELISM=8 made your changes work. Please note I'm still using the same build which does contain the commits you reverted recently in this PR.

https://monster.dragonflybsd.org/~tuxillo/grok_bench/5/opengrok0.0.log

Regards,
Antonio Huete

idodeclare commented:
@tuxillo , good result! In that run of just under eight hours, the largest repo churn was 1,508,288 files (deleted 707,178, modified 801,110) — that satisfies @vladak's request for "measurements to see if this [patch] really helps" issue #1908.

The last `forceMerge()` (which is documented in OpenGrok as "deprecated and not needed anymore") took just over three hours, so that's another goblin to examine in the future.

	2018-01-22 16:46:04.415-0800 FINE t126 PendingFileCompleter.complete: deleted 0 file(s)
	2018-01-22 16:46:04.521-0800 FINE t126 PendingFileCompleter.complete: renamed 101 file(s)
	2018-01-22 16:46:04.523-0800 FINE t126 IndexDatabase.finishWriting: completed 101 file(s)
	2018-01-22 16:47:18.011-0800 FINE t130 PendingFileCompleter.complete: deleted 0 file(s)
	2018-01-22 16:47:18.853-0800 FINE t125 PendingFileCompleter.complete: deleted 4 file(s)
	2018-01-22 16:47:20.202-0800 FINE t125 PendingFileCompleter.complete: renamed 268 file(s)
	2018-01-22 16:47:20.203-0800 FINE t125 IndexDatabase.finishWriting: completed 272 file(s)
	2018-01-22 16:47:20.234-0800 FINE t130 PendingFileCompleter.complete: renamed 5,297 file(s)
	2018-01-22 16:47:20.234-0800 FINE t130 IndexDatabase.finishWriting: completed 5,297 file(s)
	2018-01-22 16:48:33.865-0800 FINE t129 PendingFileCompleter.complete: deleted 0 file(s)
	2018-01-22 16:48:38.231-0800 FINE t129 PendingFileCompleter.complete: renamed 310 file(s)
	2018-01-22 16:48:38.232-0800 FINE t129 IndexDatabase.finishWriting: completed 310 file(s)
	2018-01-22 17:46:11.699-0800 FINE t132 PendingFileCompleter.complete: deleted 0 file(s)
	2018-01-22 17:46:12.883-0800 FINE t132 PendingFileCompleter.complete: renamed 201 file(s)
	2018-01-22 17:46:12.883-0800 FINE t132 IndexDatabase.finishWriting: completed 201 file(s)
	2018-01-22 17:49:48.044-0800 FINE t127 PendingFileCompleter.complete: deleted 1 file(s)
	2018-01-22 17:49:48.263-0800 FINE t127 PendingFileCompleter.complete: renamed 36 file(s)
	2018-01-22 17:49:48.263-0800 FINE t127 IndexDatabase.finishWriting: completed 37 file(s)
	2018-01-22 17:50:50.729-0800 FINE t131 PendingFileCompleter.complete: deleted 0 file(s)
	2018-01-22 17:50:51.636-0800 FINE t131 PendingFileCompleter.complete: renamed 399 file(s)
	2018-01-22 17:50:51.636-0800 FINE t131 IndexDatabase.finishWriting: completed 399 file(s)
	2018-01-22 17:56:58.805-0800 FINE t128 PendingFileCompleter.complete: deleted 107 file(s)
	2018-01-22 17:57:02.813-0800 FINE t128 PendingFileCompleter.complete: renamed 4,690 file(s)
	2018-01-22 17:57:02.814-0800 FINE t128 IndexDatabase.finishWriting: completed 4,797 file(s)
	2018-01-22 20:48:02.729-0800 FINE t126 PendingFileCompleter.complete: deleted 707,178 file(s)
	2018-01-22 20:53:43.904-0800 FINE t126 PendingFileCompleter.complete: renamed 801,110 file(s)
	2018-01-22 20:53:43.904-0800 FINE t126 IndexDatabase.finishWriting: completed 1,508,288 file(s)
	2018-01-22 20:53:43.904-0800 FINE t126 IndexDatabase.finishWriting: completed 1,508,288 file(s)
	2018-01-22 20:53:43.951-0800 INFO t126 IndexDatabase.optimize: Optimizing the index ... 
	2018-01-23 00:09:43.911-0800 INFO t126 IndexDatabase.optimize: done

tulinkry commented on src/org/opensolaris/opengrok/configuration/LuceneLockName.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Represents a container for OpenGrok's names of Lucene lock modes.
+ */
+public class LuceneLockName {
```
enum?

idodeclare replied:
OK revised

tulinkry commented:
This is a great job

idodeclare commented:
Thank you, @tulinkry !

tarzanek commented:
hey guys, this really looks great ... so I think we could merge and spin up an RC for this to give it a bigger testing audience ...
@vladak - can we go with the plan as written in slack (so RC for your python scripts & then RC for this?) ?

tulinkry commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -141,29 +154,31 @@ public IndexDatabase() throws IOException {
      * standard out.
      *
-     * @param executor An executor to run the job
+     * @param parallelizer a defined instance
```
this maybe a little bit more descriptive

idodeclare replied:
My reasoning behind this wording is that I think it's more useful to know whether an argument is required or not — hence "a defined instance" or "an optional instance" — rather than recapitulating the description of the class. If the method specifically targeted an argument, I might write e.g., "a defined instance to write" or "a defined instance to append". If the method did something very unintuitive, I would comment.

tulinkry replied:
ok that didn't seem obvious to me

tulinkry commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -141,29 +154,31 @@ public IndexDatabase() throws IOException {
      * Update the index database for all of the projects
      *
-     * @param executor An executor to run the job
+     * @param parallelizer a defined instance
```
this maybe a little bit more descriptive

idodeclare replied:
""

tulinkry commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -426,8 +442,15 @@ public void update() throws IOException, HistoryException {
                     }
 
                     // The actual indexing happens in indexDown().
```
so this is no longer true I guess?

tulinkry commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -483,10 +509,11 @@ public void update() throws IOException, HistoryException {
      * Optimize all index databases
      *
-     * @param executor An executor to run the job
+     * @param parallelizer a defined instance
```
this could be more descriptive

idodeclare replied:
""

tulinkry commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -945,18 +974,89 @@ private int indexDown(File dir, String parent, boolean count_only,
+    private void indexParallel(IndexDownArgs args) throws IOException {
+
+        int worksCount = args.works.size();
+        if (worksCount < 1) return;
```
some log message?

idodeclare replied:
There are already messages about number of files updated/deleted and total.

tulinkry replied:
ok

tulinkry commented on src/org/opensolaris/opengrok/index/IndexDatabase.java:
```diff
@@ -945,18 +974,89 @@ private int indexDown(File dir, String parent, boolean count_only,
+                            ret = true;
+                        } catch (InterruptedException e) {
+                            // Allow one retry if interrupted
+                            if (++tries <= 1) continue;
```
maybe a log message?

idodeclare replied:
OK added.

tulinkry commented on src/org/opensolaris/opengrok/index/IndexerParallelizer.java:
```diff
@@ -0,0 +1,143 @@
+
+    private final ExecutorService fixedExecutor;
+    private final ForkJoinPool forkJoinPool;
+    private final ScheduledThreadPoolExecutor schedExecutor;
```
unused?

idodeclare replied:
OK removed.

tulinkry commented on src/org/opensolaris/opengrok/index/PendingFileCompleter.java:
```diff
@@ -0,0 +1,409 @@
+     * specified element
+     */
+    public boolean add(PendingFileDeletion e) {
+        return deletions.add(e);
```
why this is not synchronized?

idodeclare replied:
`PendingFileCompleter` is package-private but it really might be a private static class of `IndexDatabase` and only for its size did I make it a separate file. `PendingFileCompleter` accords completely with `IndexDatabase`'s algorithm where planned deletions are determined by a single thread iterating through the filesystem while simultaneously iterating through a Lucene `TermsEnum`. Synchronizing in `add(PendingFileDeletion)` would have no benefit.

tulinkry replied:
so
```java
public boolean add(PendingFileRenaming e) {
public boolean add(PendingFileDeletion e) {
```
cant happen at the same time?

tulinkry replied:
then ok

idodeclare replied:
> cant happen at the same time?

Correct — they are only called in separate stages of processing.

N.b. for the stage where `add(PendingFileRenaming)` is being executed, within the same `INSTANCE_LOCK`, the two sets tracking deletions and renamings are both modified.

tulinkry commented:
Just to recap cause maybe I didn't see something.

This does:
1. first discover the files and queue them for add/remove
2. index the files with ctags and add it to the lucene
3. create the pending xref
4. prepare commit for the lucene index
5. execute all file deletes/moves etc (also renaming the pending xref)
6. commit the lucene index

am I sort of right?

How is the race when creating/deleting files folders in parallel solved?

So after this PR there will be two levels of parallelism:
1. Per project - per index database
2. Per file for the whole code base

Is this correct?

idodeclare commented:
> Just to recap cause maybe I didn't see something.
> ...
> 1) first discover the files and queue them for add/remove

Yes, and to clarify: at this step, affected documents are deleted from Lucene (but not committed) and xrefs deletions are enqueued. For modified source files, the documents will later be re-added back to Lucene and their xref deletions will be switched to renamings.

> am I sort of right?

Yes that is a good summary.

> How is the race when creating/deleting files folders in parallel solved?

`java.io.File` allows contention in clients' creating directories. Simultaneous clients can run `mkdirs()` for the same path where only one "succeeds" but neither throws an exception.

For removal of directories, I elected to execute that phase on a single thread. Parallelization would have required a locking mechanism on path components. E.g., threads removing `/path/to/dir/` and `/other/path/to/otherdir/` could operate simultaneously but threads removing `/path/to/dir/` and `/path/to/dir2/` would have to be synchronized because their mutual [grand]parents, `/path/to/` and `/path/`, might for cleanliness require removal too.

> So after this PR there will be two levels of parallelism:
> 
> Per project - per index database - files discovery
> Per file for the whole code base - files processing
> Is this correct?

Yes that is right. The project-level `getFixedExecutor()` threads are clients of the per-file `getForkJoinPool()` threads so that files progress at the same rate overall and large repositories get more threads as other, smaller repositories finish.

tarzanek commented:
so since we run first the discovery, I'd guess we will now have a very easy way to enable progress bar (rather than previously running the same walk first and then index ... ), but this is just a side note on stuff that is there and can be simplified(or made default) after merging this change

tulinkry commented:
Yes this progress bar is the future now

tulinkry commented:
Seems ok to me, we can merge

tarzanek commented:
let's get the party started!

tarzanek commented:
thank you @idodeclare 
let's see how this code will run

tulinkry commented:
really good job, sorry for that it took that long to merge it

idodeclare commented:
Thank you, guys! Thank you, @tuxillo, for help in testing.

tuxillo commented:
Thank you guys, you are awesome :)
Will there be another RC for this one?

vladak commented:
Yes, will spin up one today.

Dne 25. 1. 2018 19:15 napsal uživatel "Antonio Huete Jimenez" <
notifications@github.com>:

Thank you guys, you are awesome :)
Will there be another RC for this one?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/oracle/opengrok/pull/1936#issuecomment-360552661>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ACzGDIOaK_L7OEUga9ht7TnwUKdZkb7Rks5tOMS4gaJpZM4Q6Q2L>
.

vladak commented on src/org/opensolaris/opengrok/index/IndexerParallelizer.java:
```diff
@@ -0,0 +1,137 @@
+
+        int indexingParallelism = env.getIndexingParallelism();
+
+        // The order of the following is important.
```
Implicit dependencies usually hint towards incorrect programming style/patterns. The comment should explain why the ordering is important.

tuxillo commented:
Was this merged eventually?
